Kconfig option for enabling USB device.

eobalski opened the pull request:
USB_DEVICE_AUTO_ENABLE option is useful when user wants to control when to enable USB device.
By default this option is disabled and USB is enabled on system boot up.

Having enabled USB device automatically may cause loosing power events from USB device. It can happen if USB had been enabled and some events arrived but user did not register any class specific callback yet. In such scenario this information will be dropped.

Signed-off-by: Emil Obalski <emil.obalski@nordicsemi.no>

Commit:
usb: Kconfig: Add option for enabling USB device.

Auto enable option allows users to control when to enable USB device.
By default this option is disabled and USB is enabled on system boot up.

Signed-off-by: Emil Obalski <emil.obalski@nordicsemi.no>

Commit:
samples: usb: Adapt USB samples for new Kconfig option

The purpose of this commit is to adapt existing examples
for new Kconfig option USB_DEVICE_AUTO_ENABLE.
If configured 'n' then user app needs to enable usb by itself.

Signed-off-by: Emil Obalski <emil.obalski@nordicsemi.no>

carlescufi reviewed (approved):
LGTM

jfischer-phytec-iot commented on drivers/console/uart_console.c:
```diff
@@ -596,6 +599,15 @@ static int uart_console_init(struct device *arg)
 	uart_console_dev = device_get_binding(CONFIG_UART_CONSOLE_ON_DEV_NAME);
 
 #if defined(CONFIG_USB_UART_CONSOLE) && defined(CONFIG_USB_UART_DTR_WAIT)
+#ifndef CONFIG_USB_DEVICE_AUTO_ENABLE
```
To be honest, I would not expect `usb_enable()` here. We probably have to add a mutex to `usb_enable()`.

```diff
diff --git a/subsys/usb/usb_device.c b/subsys/usb/usb_device.c
index 96727c5af3..339cf9521d 100644
--- a/subsys/usb/usb_device.c
+++ b/subsys/usb/usb_device.c
@@ -110,6 +110,8 @@ LOG_MODULE_REGISTER(usb_device);
 extern struct usb_cfg_data __usb_data_start[];
 extern struct usb_cfg_data __usb_data_end[];
 
+K_MUTEX_DEFINE(usb_lock);
+
 struct usb_transfer_data {
        /** endpoint associated to the transfer */
        u8_t ep;
@@ -1542,10 +1544,12 @@ int usb_enable(void)
                return 0;
        }
 
+       k_mutex_lock(&usb_lock, K_FOREVER);
+
        /* Enable VBUS if needed */
        ret = usb_vbus_set(true);
        if (ret < 0) {
-               return ret;
+               goto usb_enable_error;
        }
 
        usb_register_status_callback(forward_status_cb);
@@ -1553,7 +1557,7 @@ int usb_enable(void)
 
        ret = usb_dc_attach();
        if (ret < 0) {
-               return ret;
+               goto usb_enable_error;
        }
 
        /* Configure control EP */
@@ -1563,32 +1567,32 @@ int usb_enable(void)
        ep0_cfg.ep_addr = USB_CONTROL_OUT_EP0;
        ret = usb_dc_ep_configure(&ep0_cfg);
        if (ret < 0) {
-               return ret;
+               goto usb_enable_error;
        }
 
        ep0_cfg.ep_addr = USB_CONTROL_IN_EP0;
        ret = usb_dc_ep_configure(&ep0_cfg);
        if (ret < 0) {
-               return ret;
+               goto usb_enable_error;
        }
        /* Register endpoint 0 handlers*/
        ret = usb_dc_ep_set_callback(USB_CONTROL_OUT_EP0,
                                     usb_handle_control_transfer);
        if (ret < 0) {
-               return ret;
+               goto usb_enable_error;
        }
 
        ret = usb_dc_ep_set_callback(USB_CONTROL_IN_EP0,
                                     usb_handle_control_transfer);
        if (ret < 0) {
-               return ret;
+               goto usb_enable_error;
        }
 
        /* Register endpoint handlers*/
        ret = composite_setup_ep_cb();
        if (ret < 0) {
-               return ret;
+               goto usb_enable_error;
        }
 
        /* Init transfer slots */
@@ -1600,17 +1604,20 @@ int usb_enable(void)
        /* Enable control EP */
        ret = usb_dc_ep_enable(USB_CONTROL_OUT_EP0);
        if (ret < 0) {
-               return ret;
+               goto usb_enable_error;
        }
 
        ret = usb_dc_ep_enable(USB_CONTROL_IN_EP0);
        if (ret < 0) {
-               return ret;
+               goto usb_enable_error;
        }
 
        usb_dev.enabled = true;
 
-       return 0;
+usb_enable_error:
+       k_mutex_unlock(&usb_lock);
+
+       return ret;
 }
 
 /*

```

emob-nordic replied:
I actually do not get what's the point for that. By the Kconfig option and #ifndefs I assume the usb_enable will be called once. Is mutex really necessary in this case?

jfischer-phytec-iot replied:
For example a application has 3 interfaces: 2 x CDC ACM, one of them for console; 1 x MSC

CONFIG_USB_COMPOSITE_DEVICE=y
CONFIG_USB_CDC_ACM=y
CONFIG_USB_CDC_ACM_DEVICE_COUNT=2
...
CONFIG_USB_UART_CONSOLE=y
CONFIG_USB_DEVICE_AUTO_ENABLE=n
CONFIG_USB_MASS_STORAGE=y

Is it safe that usb_enable() can not be interrupted and than be called by another thread?

emob-nordic replied:
Okay, now I see your point and ofc You are right. I will introduce mutex here.

emob-nordic commented on drivers/console/uart_console.c:
```diff
@@ -596,6 +599,15 @@ static int uart_console_init(struct device *arg)
+
+	ret = usb_enable();
+	if (ret < 0) {
+		printk("Failed to enable USB device");
```
Ah, You are right. This will not appear anywhere. In this case I will remove this error checking.

jfischer-phytec-iot replied:
just remove `printk("Failed to enable USB device");`

emob-nordic commented on samples/subsys/usb/cdc_acm/src/main.c:
```diff
@@ -19,6 +19,9 @@
 #include <sys/ring_buffer.h>
 
+#ifndef CONFIG_USB_DEVICE_AUTO_ENABLE
+#include <usb/usb_device.h>
```
Do You refer to all samples here or just this particular one?

jfischer-phytec-iot replied:
To all, but I do not insist. Leave it as you did.

emob-nordic commented on drivers/console/uart_console.c:
```diff
@@ -33,6 +33,9 @@
 #ifdef CONFIG_UART_CONSOLE_MCUMGR
 #include "mgmt/serial.h"
 #endif
+#ifdef CONFIG_USB_DEVICE_AUTO_ENABLE
```
This should be ifndef. I will force push that.

Commit:
usb: Add mutex to usb_enable

The purpose of this commit is to prevent user from calling
usb_enable() from different contexes.

If for example user configures USB_DEVICE_AUTO_ENABLE=n
and composite device with 2 cdc_acm classes running on
different threads this mechanism will prevent from calling
usb_enable twice and thus lead to undefined behaviour of
usb driver controller.

usb_enable shall always be called once as it refers to usb driver
and for now only one USB driver instance is supported in Zephyr.
This mechanism ensures that.

Signed-off-by: Emil Obalski <emil.obalski@nordicsemi.no>

jfischer-no reviewed (changes requested):
Can you also update _samples/bluetooth/hci_usb, samples/boards/intel_s1000_crb, samples/net/wpanusb, samples/sensor/fxos8700-hid_, please?

emob-nordic commented:
> Can you also update _samples/bluetooth/hci_usb, samples/boards/intel_s1000_crb, samples/net/wpanusb, samples/sensor/fxos8700-hid_, please?

Will be done in #20375
